packages/workflow/src/NodeErrors.ts

```diff
@@ -132,9 +141,33 @@ abstract class NodeError extends Error {
 	/**
 	 * Check if a value is an object with at least one key, i.e. it can be traversed.
 	 */
-	private isTraversableObject(value: any): value is IRawErrorObject { // tslint:disable-line:no-any
+	protected isTraversableObject(value: any): value is JsonObject { // tslint:disable-line:no-any
 		return value && typeof value === 'object' && !Array.isArray(value) && !!Object.keys(value).length;
 	}
+
+	/**
+	 * Remove circular references from objects.
+	 */
+	 protected removeCircularRefs(obj: JsonObject, seen = new Set()) {
+		seen.add(obj);
+		Object.entries(obj).forEach(([key, value]) => {
+			if (this.isTraversableObject(value)) {
+				seen.has(value) ? obj[key] = '[circular]' : this.removeCircularRefs(value, seen);
+				return;
+			}
+			if (Array.isArray(value)) {
+				value.forEach((val, index) => {
+					if (seen.has(val)) {
+						value[index] = '[circular]';
+						return;
+					}
+					if (this.isTraversableObject(val)) {
+						this.removeCircularRefs(val, seen);
+					}
+				});
+			}
```

Comment:
we don't handle nested arrays here, do we keep it out of scope to not further add complexity? @ivov

Reply:
it's not outdated 🙈

Reply:
decided to keep it out of scope for now